File: securerag/builtin_schemes.py
from __future__ import annotations

import hashlib
import hmac
import re
import secrets
from typing import Any

from securerag.scheme_plugin import EncryptedSchemePlugin


def _tokenize(text: str) -> list[str]:
    raw = re.findall(r"[a-z0-9]+", text.lower())
    out = []
    for tok in raw:
        if len(tok) > 3 and tok.endswith("s"):
            tok = tok[:-1]
        out.append(tok)
    return out


def _encrypt_token(token: str, key: str) -> str:
    return hmac.new(key.encode(), token.encode(), hashlib.sha256).hexdigest()


class SSEPlugin(EncryptedSchemePlugin):
    def generate_key(self) -> str:
        return secrets.token_hex(16)

    def prepare_chunk(self, text: str, key: str) -> dict[str, Any]:
        return {"enc_terms": [_encrypt_token(t, key) for t in _tokenize(text)]}

    def build_server_index(self, rows: list[dict[str, Any]]) -> dict[str, Any]:
        inv: dict[str, list[int]] = {}
        for i, row in enumerate(rows):
            for term in row.get("scheme_data", {}).get("enc_terms", []):
                inv.setdefault(term, []).append(i)
        return {"rows": rows, "inv": inv}

    def encrypt_query(self, query: str, key: str) -> dict[str, Any]:
        return {"enc_terms": [_encrypt_token(t, key) for t in _tokenize(query)]}

    def search(
        self,
        server_index: Any,
        encrypted_query: dict[str, Any],
        top_k: int,
    ) -> list[dict[str, Any]]:
        rows = server_index["rows"]
        inv = server_index["inv"]
        q_terms = encrypted_query.get("enc_terms", [])
        if not q_terms:
            return []

        counts: dict[int, int] = {}
        for term in q_terms:
            for idx in inv.get(term, []):
                counts[idx] = counts.get(idx, 0) + 1

        q_len = len(q_terms)
        scored = []
        for idx, inter in counts.items():
            row = rows[idx]
            doc_terms = row.get("scheme_data", {}).get("enc_terms", [])
            union = q_len + len(doc_terms) - inter
            score = inter / union if union else 0.0
            scored.append(
                {
                    "doc_id": row["doc_id"],
                    "text": row["text"],
                    "metadata": row.get("metadata", {}),
                    "score": score,
                }
            )
        scored.sort(key=lambda x: x["score"], reverse=True)
        return scored[:top_k]


class StructuredPlugin(EncryptedSchemePlugin):
    def __init__(self, use_bigrams: bool = True):
        self.use_bigrams = use_bigrams

    def generate_key(self) -> str:
        return secrets.token_hex(16)

    def _make_terms(self, text: str, key: str) -> list[str]:
        tokens = _tokenize(text)
        out = [_encrypt_token(f"tok:{t}", key) for t in tokens]
        if self.use_bigrams and len(tokens) >= 2:
            for i in range(len(tokens) - 1):
                out.append(_encrypt_token(f"bi:{tokens[i]}_{tokens[i + 1]}", key))
        return out

    def prepare_chunk(self, text: str, key: str) -> dict[str, Any]:
        return {"struct_terms": self._make_terms(text, key)}

    def build_server_index(self, rows: list[dict[str, Any]]) -> dict[str, Any]:
        inv: dict[str, list[int]] = {}
        for i, row in enumerate(rows):
            for term in row.get("scheme_data", {}).get("struct_terms", []):
                inv.setdefault(term, []).append(i)
        return {"rows": rows, "inv": inv}

    def encrypt_query(self, query: str, key: str) -> dict[str, Any]:
        return {"struct_terms": self._make_terms(query, key)}

    def search(
        self,
        server_index: Any,
        encrypted_query: dict[str, Any],
        top_k: int,
    ) -> list[dict[str, Any]]:
        rows = server_index["rows"]
        inv = server_index["inv"]
        q_terms = encrypted_query.get("struct_terms", [])
        if not q_terms:
            return []

        counts: dict[int, int] = {}
        for term in q_terms:
            for idx in inv.get(term, []):
                counts[idx] = counts.get(idx, 0) + 1

        q_len = len(q_terms)
        scored = []
        for idx, inter in counts.items():
            row = rows[idx]
            doc_terms = row.get("scheme_data", {}).get("struct_terms", [])
            union = q_len + len(doc_terms) - inter
            score = inter / union if union else 0.0
            scored.append(
                {
                    "doc_id": row["doc_id"],
                    "text": row["text"],
                    "metadata": row.get("metadata", {}),
                    "score": score,
                }
            )
        scored.sort(key=lambda x: x["score"], reverse=True)
        return scored[:top_k]


EncryptedSchemePlugin.register("sse", SSEPlugin())
EncryptedSchemePlugin.register("structured", StructuredPlugin(use_bigrams=True))
EncryptedSchemePlugin.register("structured_encryption", StructuredPlugin(use_bigrams=True))

File: securerag-rs/src/builtin_schemes.rs
use std::collections::HashMap;

#[cfg(test)]
use hmac::{Hmac, Mac};
#[cfg(test)]
use sha2::Sha256;

use crate::encrypted_scheme::{EncryptedScheme, SchemeIndex, SchemeRow, SearchResult};

pub struct RustSSEScheme;

pub struct RustStructuredScheme {
    pub use_bigrams: bool,
}

struct InvertedIndex {
    rows: Vec<(String, String, prost_types::Struct, Vec<String>)>,
    inv: HashMap<String, Vec<usize>>,
    query_field: &'static str,
}

impl InvertedIndex {
    fn build(rows: &[SchemeRow], field: &'static str, query_field: &'static str) -> Self {
        let mut packed_rows: Vec<(String, String, prost_types::Struct, Vec<String>)> =
            Vec::with_capacity(rows.len());
        let mut inv: HashMap<String, Vec<usize>> = HashMap::new();
        for (i, row) in rows.iter().enumerate() {
            let terms = row
                .scheme_data
                .get(field)
                .and_then(|v| v.as_array())
                .map(|a| {
                    a.iter()
                        .filter_map(|x| x.as_str().map(str::to_string))
                        .collect::<Vec<_>>()
                })
                .unwrap_or_default();
            for term in &terms {
                inv.entry(term.clone()).or_default().push(i);
            }
            packed_rows.push((
                row.doc_id.clone(),
                row.text.clone(),
                row.metadata.clone(),
                terms,
            ));
        }
        Self {
            rows: packed_rows,
            inv,
            query_field,
        }
    }
}

impl SchemeIndex for InvertedIndex {
    fn search(
        &self,
        encrypted_query: &HashMap<String, serde_json::Value>,
        top_k: usize,
    ) -> Vec<SearchResult> {
        let q_terms = encrypted_query
            .get(self.query_field)
            .and_then(|v| v.as_array())
            .map(|a| {
                a.iter()
                    .filter_map(|x| x.as_str().map(str::to_string))
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();
        if q_terms.is_empty() {
            return Vec::new();
        }

        let mut counts: HashMap<usize, usize> = HashMap::new();
        for term in &q_terms {
            if let Some(postings) = self.inv.get(term) {
                for idx in postings {
                    *counts.entry(*idx).or_insert(0) += 1;
                }
            }
        }

        let q_len = q_terms.len();
        let mut scored: Vec<SearchResult> = counts
            .into_iter()
            .map(|(idx, inter)| {
                let row = &self.rows[idx];
                let union = q_len + row.3.len() - inter;
                let score = if union == 0 {
                    0.0
                } else {
                    inter as f64 / union as f64
                };
                SearchResult {
                    doc_id: row.0.clone(),
                    text: row.1.clone(),
                    metadata: row.2.clone(),
                    score,
                }
            })
            .collect();

        scored.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
        scored.truncate(top_k);
        scored
    }
}

impl EncryptedScheme for RustSSEScheme {
    fn build_index(&self, rows: &[SchemeRow]) -> Box<dyn SchemeIndex> {
        Box::new(InvertedIndex::build(rows, "enc_terms", "enc_terms"))
    }
}

impl EncryptedScheme for RustStructuredScheme {
    fn build_index(&self, rows: &[SchemeRow]) -> Box<dyn SchemeIndex> {
        let _ = self.use_bigrams;
        Box::new(InvertedIndex::build(rows, "struct_terms", "struct_terms"))
    }
}

#[cfg(test)]
pub fn hmac_sha256_hex(token: &str, key: &str) -> String {
    let mut mac = Hmac::<Sha256>::new_from_slice(key.as_bytes()).expect("valid key");
    mac.update(token.as_bytes());
    let out = mac.finalize().into_bytes();
    out.iter().map(|b| format!("{:02x}", b)).collect()
}

#[cfg(test)]
mod tests {
    use super::hmac_sha256_hex;

    #[test]
    fn hmac_hex_parity_contract() {
        assert_eq!(
            hmac_sha256_hex("tok:alpha", "abcd1234abcd1234abcd1234abcd1234"),
            "b5187317422e4f4b011ba33c1450032a763c9337170feb04932b94ecbcc21faa"
        );
    }
}
